aten/src/ATen/native/sparse/cuda/SparseMatMul.cu

```diff
@@ -0,0 +1,884 @@
+#include <ATen/ATen.h>
+#include <ATen/Config.h>
+#include <ATen/NamedTensorUtils.h>
+#include <ATen/NativeFunctions.h>
+#include <ATen/Parallel.h>
+#include <ATen/SparseTensorImpl.h>
+#include <ATen/SparseTensorUtils.h>
+#include <cuda_runtime.h>
+#include <cusparse_v2.h>
+
+#include <thrust/device_ptr.h>
+#include <thrust/for_each.h>
+#include <thrust/sequence.h>
+
+#include <THC/THCTensorMathPointwise.cuh>
+#include <THC/THCThrustAllocator.cuh>
+
+#include <ATen/cuda/CUDAContext.h>
+#include <ATen/cuda/CUDAUtils.h>
+
+#include <ATen/native/sparse/cuda/SparseCUDABlas.cuh>
+
+#include <thrust/device_vector.h>
+#include <thrust/host_vector.h>
+
+#if defined(__CUDACC__) && (CUSPARSE_VERSION >= 11000)
+#define IS_CUSPARSE11_AVAILABLE() 1
+#else
+#define IS_CUSPARSE11_AVAILABLE() 0
+#endif
+
+#if IS_CUSPARSE11_AVAILABLE()
+#include <library_types.h>
+#endif
+
+namespace at {
+namespace native {
+
+namespace {
+
+using namespace at::sparse;
+
+IntTensor _to_csr_int(const LongTensor& rowIndices, int64_t dim, int64_t nnz) {
+  IntTensor csr = at::empty({dim + 1}, CUDA(kInt));
+  IntTensor rowIndicesInt = at::empty({rowIndices.size(0)}, CUDA(kInt));
+  rowIndicesInt.copy_(rowIndices);
+  sparse::cuda::Xcoo2csr(
+      rowIndicesInt.data_ptr<int32_t>(), nnz, dim, csr.data_ptr<int32_t>());
+  return csr;
+}
+
+int confirm_mult_size(const std::vector<int>& mat1_size, const std::vector<int>& mat2_size) {
+  TORCH_CHECK(
+      mat1_size[1] == mat2_size[0],
+      "mat1 and mat2 shapes cannot be multiplied (",
+      mat1_size[0],
+      "x",
+      mat1_size[1],
+      " and ",
+      mat2_size[0],
+      "x",
+      mat2_size[1],
+      ")");
+  return mat1_size[1];
+}
+
+void create_general_description_(cusparseMatDescr_t& description_) {
+  TORCH_CUDASPARSE_CHECK(cusparseCreateMatDescr(&description_));
```

Comment:
At this point, I'd expect a call to `cusparseDestroyMatDescr`, but I don't see any occurrences of this in this file.

More generally, it is probably less error prone if you make some C++ helper classes for managing the lifetime of these descriptors. An example of us doing it elsewhere is in aten/src/ATen/cudnn/Descriptors.h; you probably want something analogous for cusparse.